config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/PolicyEntityBuilder.java

```diff
@@ -178,6 +181,57 @@ static void prepareEncapsulatedAssertion(Policy policy, Bundle bundle, Document
         }
     }
 
+    @VisibleForTesting
+    static void prepareAuthenticationAssertion(Policy policy, Bundle bundle, Document policyDocument, Element authenticationAssertionElement) {
+        final Element idProvNameElement = getSingleChildElement(authenticationAssertionElement, ID_PROV_NAME);
+        if (idProvNameElement.hasAttribute(STRING_VALUE)) {
+            final String name = idProvNameElement.getAttribute(STRING_VALUE);
+            final String id = findIdentityProviderId(policy, bundle, name);
+            updateAuthenticationAssertion(policyDocument, authenticationAssertionElement, idProvNameElement, id);
+        } else {
+            LOGGER.log(Level.FINE, "No name specified for id provider in policy: \"{0}\".", policy.getPath());
+        }
+    }
+
+    private static void updateAuthenticationAssertion(Document policyDocument, Element authenticationAssertionElement, Element nameElement, String id) {
+        Element idProvOidElement = createElementWithAttribute(
+                policyDocument,
+                ID_PROV_OID,
+                GOID_VALUE,
+                id
+        );
+        Node firstChild = authenticationAssertionElement.getFirstChild();
+        if (firstChild != null) {
+            authenticationAssertionElement.insertBefore(idProvOidElement, firstChild);
+        } else {
+            authenticationAssertionElement.appendChild(idProvOidElement);
+        }
+
+        authenticationAssertionElement.removeChild(nameElement);
+    }
+
+    private static String findIdentityProviderId(Policy policy, Bundle bundle, String name) {
+        LOGGER.log(Level.FINE, "Looking for referenced identity provider: {0}", name);
+        if (name.equals(INTERNAL_IDP_NAME)) {
+            return INTERNAL_IDP_ID;
+        }
+        final AtomicReference<IdentityProvider> referenceIdentityProvider = new AtomicReference<>(bundle.getIdentityProviders().get(name));
```

Comment:
This will need to be handled differently because we won't know the ID provider name at deployment bundle build time. We will only know the ID at Environment bundle creation time. It should work somewhat similarly to the way the ENV. context variables are set. In the [PolicyEntityBuilder we put a reference](https://github.com/ca-api-gateway/gateway-developer-plugin/blob/9136ede3a6c02244fb1eeb93f8d7d5951cfa6af2/config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/PolicyEntityBuilder.java#L124). And in the [BundleDetemplatizer we provide the value](https://github.com/ca-api-gateway/gateway-developer-plugin/blob/9136ede3a6c02244fb1eeb93f8d7d5951cfa6af2/environment-creator-application/src/main/java/com/ca/apim/gateway/cagatewayconfig/BundleDetemplatizer.java#L28).

Reply:
I will look into referencing and using the bundleDetemplatizer way, thanks for the quick links